### DISCUSSION TITLE
Error message from ‘msc313-fcie’

### DISCUSSION THREAD

**Discussion by ltrsunny** [linux-chenxing/linux-chenxing.org#50] [General]
Build from branch ‘mstar_v5_15_rebase’, working with a low CPU & memory consumption but very slow. 
Got multiple times of such messages:
```
[   38.888642] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[   38.888668] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[   38.895767] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[   70.248644] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[   70.248673] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[   70.255773] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
```

**Comment by fifteenhex** [comment]
The driver might need some tweaks. Does the SD card keep working when you see those messages?

**Reply by ltrsunny** [reply]
Should be working because my rootfs is on SD. But some times does not respond like 'typed in and then characters appeared aftter few seconds'.

**Reply by ltrsunny** [reply]
And the messages do not appear regularly, some times only in booting, sometimes keep notifying

**Reply by fifteenhex** [reply]
I see those messages when the kernel initially tries to probe the card and it doesn't respond to the commands the kernel is trying... I think you might be just seeing the card ignoring something the kernel is trying to do. It might be harmless.

Tell me if you get issues like the card locking up and we'll try to debug what is going on.

**Reply by ltrsunny** [reply]
Ok let me try to do more test.

**Reply by ltrsunny** [reply]
It happens while exact I/O operations on SD. When it is reporting the errors, the whole system becomes unusable slow (I cannot even type in via uart):
```
[   28.148239] ext4 filesystem being remounted at /run/systemd/unit-root/var/lib/systemd/timesync supports timestamps until 2038 (0x7fffffff)
[   28.165949] ext4 filesystem being remounted at /run/systemd/unit-root/var/tmp supports timestamps until 2038 (0x7fffffff)
[   32.487328] vcc_dram: disabling
[   79.847335] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[   79.856979] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[   79.864082] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[  699.369154] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[  699.378799] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[  699.385915] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[  737.129341] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[  737.139001] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[  737.146124] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[  767.849463] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[  767.859131] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[  767.866266] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[  799.209574] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[  799.219243] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[  799.226361] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[  849.129629] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[  849.139300] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[  849.146445] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[  888.169641] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[  888.179309] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[  888.186453] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[  934.889685] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[  934.899354] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[  934.906480] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[  966.889641] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[  966.899311] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[  966.906434] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[ 1000.819616] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[ 1000.829286] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[ 1000.836413] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[ 1061.609583] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[ 1061.619237] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[ 1061.626355] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[ 1100.649545] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[ 1100.659200] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[ 1100.666320] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[ 1154.409487] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[ 1154.419140] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[ 1154.426269] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[ 1186.409444] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[ 1186.419114] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[ 1186.426238] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[ 1243.369008] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[ 1243.378662] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[ 1243.385778] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
[ 1341.288169] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[ 1341.297822] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[ 1341.304947] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
```
from 699, I am running a 'apt get'.

I am using swap on SD, I guess this is why it appears irregularly in normal using.

**Reply by ltrsunny** [reply]
Any idea Daniel? @fifteenhex

**Reply by fifteenhex** [reply]
I found a similar issue in u-boot and the cause was just not waiting long enough for the transfer to finish.

Maybe you could enable the mmc tracing in the kernel and dump out the commands before the messages?

**Reply by ltrsunny** [reply]
> I found a similar issue in u-boot and the cause was just not waiting long enough for the transfer to finish.
> 
> Maybe you could enable the mmc tracing in the kernel and dump out the commands before the messages?

I enabled `MMC_DEBUG` and `FAIL_MMC_REQUEST` but there are still only 3 same messages repeating
```

[  413.305491] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
[  413.315137] msc313-fcie 1f282000.sdio: timedout and no status flags were set
[  413.322249] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000

```

**Reply by ltrsunny** [reply]
> I enabled `MMC_DEBUG` and `FAIL_MMC_REQUEST` but there are still only 3 same messages repeating

Should I enable more? @fifteenhex

**Reply by fifteenhex** [reply]
mmm let me try to debug this. I'm getting the same issue chroot'd into a debian system and running apt-get update.
When I ran debian before I did it with a USB drive.

**Reply by fifteenhex** [reply]
Try changing ```#define FCIE_CMD_TIMEOUT_MS	30000``` at the top of `drivers/mmc/host/mstar-fcie.c` to ```#define FCIE_CMD_TIMEOUT_MS	250```. You will still get the messages but it'll stop the driver from locking everything up for 30s.

As far as I can tell data is getting written correctly. We are just either not getting a response from the card or the hardware isn't sending the command in some cases.

I might need to wire this up to my logic analyzer to confirm which it is but I need a better way to reproduce it for that to be useful.

**Reply by ltrsunny** [reply]
Thanks for your solution.

**Reply by ltrsunny** [reply]
> Try changing `#define FCIE_CMD_TIMEOUT_MS 30000` at the top of `drivers/mmc/host/mstar-fcie.c` to `#define FCIE_CMD_TIMEOUT_MS 250`. You will still get the messages but it'll stop the driver from locking everything up for 30s.
> 
> As far as I can tell data is getting written correctly. We are just either not getting a response from the card or the hardware isn't sending the command in some cases.
> 
> I might need to wire this up to my logic analyzer to confirm which it is but I need a better way to reproduce it for that to be useful.

Seem getting better but far from ‘usable’ I think.

**Reply by fifteenhex** [reply]
> Seem getting better but far from ‘usable’ I think.

yep, I think I know where the issue is now (The command is timing out because the hardware isn't running it) but I'm not sure why it happens.

**Reply by ltrsunny** [reply]
> > Seem getting better but far from ‘usable’ I think.
> 
> yep, I think I know where the issue is now (The command is timing out because the hardware isn't running it) but I'm not sure why it happens.

Did you get TAKOYAKI_DLC00V050.20211216 which just released by Sigmastar? Perhaps it can be a reference.

Anyway, happy new year Daniel.

**Reply by fifteenhex** [reply]
I don't have the new kernel from that SDK yet. I don't have an NDA with sigmastar so I have to wait until someone uploads the kernel source.

I think the problem is sometimes the card isn't finished doing what it's doing between writing a block of data and stopping the transfer.
I've made it better but I can't make it go away totally. Hopefully we'll get some documentation for the SSD202D soon. I think we are just one setting bit away from being reliable.

**Reply by ltrsunny** [reply]
It doesn't need NDA and already uploaded to Baidu disk. I can send the link to you.

**Reply by fifteenhex** [reply]
Thanks.

I will keep looking at this. I need to study the SD card protocol again to make sure my code is right :(

**Reply by ltrsunny** [reply]
> Thanks.
> 
> I will keep looking at this. I need to study the SD card protocol again to make sure my code is right :(

Maybe it is not about the driver itself, but the kernel's method of using the driver has been changed. 
I am also trying to read the code but may not help much in this issue because my experience is mainly focused on circuit design.

**Reply by fifteenhex** [reply]
> > Thanks.
> > I will keep looking at this. I need to study the SD card protocol again to make sure my code is right :(
> 
> Maybe it is not about the driver itself, but the kernel's method of using the driver has been changed. I am also trying to read the code but may not help much in this issue because my experience is mainly focused on circuit design.

Yes, that is also possible. The commands used by the latest kernel might be different to the old Sigmastar kernel.
If I could find an SD card protocol analyzer it would be much easier to see what the hardware is doing.

**Reply by fifteenhex** [reply]
I think I've *fixed* this now.

We were doing CMD25 to start a multiblock write and terminating with CMD12.
Apparently this termination with CMD12 is prone to timing issues.
So I've told the kernel to use CMD23 before CMD25 to set the amount of blocks being written and now we don't need to use CMD12.. and the warnings are gone.
I think for CMD12 to work properly it needs be to done in the IRQ handler when the CMD25 has completed so there is no delay in sending it..

**Reply by fifteenhex** [reply]
Please try with mstar_v5_16_rebase now (you'll probably need to force pull). Most of the issues should be gone now. It needs some fixes to make it faster at some point.

**Reply by ltrsunny** [reply]
Cool! Let me try later.

**Reply by ltrsunny** [reply]
Have been testing a whole day. Much better than before but still sometimes frozen. I am not sure whether it is about MMC now.

**Reply by fifteenhex** [reply]
Does it lock up entirely forever or does it eventually come back again?

**Reply by ltrsunny** [reply]
> Does it lock up entirely forever or does it eventually come back again?

It will recover ~~and there is nothing new in `dmesg` now~~
There is still something wrong in initialization
```
[    1.528477] gpio gpiochip1: (gpio-msc313-pm): failed to allocate parent hwirq 73 for hwirq 1
[    1.537633] msc313-fcie 1f282000.sdio: err during job; status: 0008, ctrl: 0206, blksz: 0200, blkcnt: 0001, cmdrspsz: 0505
[    1.548830] msc313-fcie 1f282000.sdio: err during job; cmd 1 (1), data 0 (0), busy 0 (0)
[    1.556988] msc313-fcie 1f282000.sdio: no response from card, removed?
[    1.563592] msc313-fcie 1f282000.sdio: failed to send command; cmd: 52 arg: 0x00000c00
[    1.572159] msc313-fcie 1f282000.sdio: err during job; status: 0008, ctrl: 0206, blksz: 0200, blkcnt: 0001, cmdrspsz: 0505
[    1.583364] msc313-fcie 1f282000.sdio: err during job; cmd 1 (1), data 0 (0), busy 0 (0)
[    1.589184] msc313-rsa 1f224480.rsa: in: fffefdfcfbfaf9f8f7f6f5f4f3f2f1f0efeeedecebeae9e8e7e6e5e4e3e2e1e0dfdedddcdbdad9d8d7d6d5d4d3d2d1d0cfcecdcccbcac9c8c7c6c5c4c3c2c1c0
[    1.591534] msc313-fcie 1f282000.sdio: no response from card, removed?
[    1.606802] msc313-rsa 1f224480.rsa: out: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
[    1.613317] msc313-fcie 1f282000.sdio: failed to send command; cmd: 52 arg: 0x80000c08
[    1.629028] cpu cpu0: Looking up cpu-supply from device tree
[    1.636815] cpu cpu0: Looking up cpu-supply from device tree
[    1.637564] clk: Not disabling unused clocks
[    1.641950] ALSA device list:
[    1.644943]   No soundcards found.
[    1.648808] Waiting for root device /dev/mmcblk0p1...
[    1.656109] msc313-fcie 1f282000.sdio: no response from card, removed?
[    1.662730] msc313-fcie 1f282000.sdio: failed to send command; cmd: 5 arg: 0x00000000
[    1.671124] msc313-fcie 1f282000.sdio: no response from card, removed?
[    1.677693] msc313-fcie 1f282000.sdio: failed to send command; cmd: 5 arg: 0x00000000
[    1.686110] msc313-fcie 1f282000.sdio: no response from card, removed?
[    1.692722] msc313-fcie 1f282000.sdio: failed to send command; cmd: 5 arg: 0x00000000
[    1.701120] msc313-fcie 1f282000.sdio: no response from card, removed?
[    1.707688] msc313-fcie 1f282000.sdio: failed to send command; cmd: 5 arg: 0x00000000
[    1.743301] mmc0: host does not support reading read-only switch, assuming write-enable
[    1.753826] mmc0: new high speed SDHC card at address 1234
[    1.760028] mmcblk0: mmc0:1234 SA16G 14.5 GiB
[    1.766032]  mmcblk0: p1
```

**Comment by fifteenhex** [comment]
```
[ 2983.533705] mmc0: starting CMD25 arg 000f4ac0 flags 000000b5
[ 2983.533720] mmc0:     blksz 512 blocks 8 flags 00000100 tsac 3000 ms nsac 0
[ 2983.533736] mmc0:     CMD12 arg 00000000 flags 0000049d
[ 2983.535481] mmc0: req done (CMD25): 0: 00000900 00000000 00000000 00000000
[ 2983.535504] mmc0:     4096 bytes transferred: 0
[ 2983.535515] mmc0:     (CMD12): 0: 00000c00 00000000 00000000 00000000
[ 2983.535548] mmc0: starting CMD13 arg 59b40000 flags 00000195
[ 2983.535608] mmc0: req done (CMD13): 0: 00000900 00000000 00000000 00000000
[ 2983.535637] mmc0: starting CMD25 arg 000f4ac8 flags 000000b5
[ 2983.535653] mmc0:     blksz 512 blocks 8 flags 00000100 tsac 3000 ms nsac 0
[ 2983.535669] mmc0:     CMD12 arg 00000000 flags 0000049d
[ 2983.537441] mmc0: req done (CMD25): 0: 00000900 00000000 00000000 00000000
[ 2983.537468] mmc0:     4096 bytes transferred: 0
[ 2[ 3014.166621] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
983.537478] mmc0[ 3014.176547] msc313-fcie 1f282000.sdio: timedout and no status flags were set
:     (CMD12): 0[ 3014.185006] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
: 00000c00 00000000 00000000 00000000
```

**Reply by fifteenhex** [reply]
Looks like sometimes the response for CMD12 (stop) after a CMD25 (multiblock transfer) for dirty page writeout isn't coming or isn't triggering an interrupt. The data is getting written correctly though.

**Reply by ltrsunny** [reply]
> ```
> [ 2983.533705] mmc0: starting CMD25 arg 000f4ac0 flags 000000b5
> [ 2983.533720] mmc0:     blksz 512 blocks 8 flags 00000100 tsac 3000 ms nsac 0
> [ 2983.533736] mmc0:     CMD12 arg 00000000 flags 0000049d
> [ 2983.535481] mmc0: req done (CMD25): 0: 00000900 00000000 00000000 00000000
> [ 2983.535504] mmc0:     4096 bytes transferred: 0
> [ 2983.535515] mmc0:     (CMD12): 0: 00000c00 00000000 00000000 00000000
> [ 2983.535548] mmc0: starting CMD13 arg 59b40000 flags 00000195
> [ 2983.535608] mmc0: req done (CMD13): 0: 00000900 00000000 00000000 00000000
> [ 2983.535637] mmc0: starting CMD25 arg 000f4ac8 flags 000000b5
> [ 2983.535653] mmc0:     blksz 512 blocks 8 flags 00000100 tsac 3000 ms nsac 0
> [ 2983.535669] mmc0:     CMD12 arg 00000000 flags 0000049d
> [ 2983.537441] mmc0: req done (CMD25): 0: 00000900 00000000 00000000 00000000
> [ 2983.537468] mmc0:     4096 bytes transferred: 0
> [ 2[ 3014.166621] msc313-fcie 1f282000.sdio: timeout waiting for interrupt, int: 0000, status: 0000, ctrl: 0306
> 983.537478] mmc0[ 3014.176547] msc313-fcie 1f282000.sdio: timedout and no status flags were set
> :     (CMD12): 0[ 3014.185006] msc313-fcie 1f282000.sdio: stop command timeout; cmd: 0x0c arg: 0x00000000
> : 00000c00 00000000 00000000 00000000
> ```

How do you get these outputs?

**Reply by fifteenhex** [reply]
I used dynamic debug to enable the debug messages..

**Comment by ltrsunny** [comment]
Any ideas about the [bug](https://github.com/linux-chenxing/linux/issues/2)？
